File: console/wechat.jl
using RedisAlchemy
using OhMyJulia
using PyCall
using JSON
using Fire

@pyimport itchat

const conn = RedisConnectionPool(db=1)
const msgq = RedisCondition{String}(conn, "channel/wechat.msg")

alive = true

@schedule while true # heartbeating
    alive && exec(conn, "set", "status/wechat.alive", "", "EX", "25")
    sleep(2)
end

@schedule while true # forward msg
    sleep(1)
end

@main function main()
    itchat.auto_login(hotReload=true, statusStorageDir="/tmp/piggy.wechat.pkl")
    itchat.run(blockThread=false)
    alive = true

    subscriber = itchat.search_chatrooms("coineering")[]["UserName"]

    while true
        msg = wait(msgq)
        itchat.send_msg(msg, subscriber)
    end
end

File: console/wechat.py
import itchat
import redis
import time

@itchat.msg_register(itchat.content.TEXT, isFriendChat=True, isGroupChat=True, isMpChat=True)
def echo(msg):
    if "alive" in msg['Text']:
        return "I'm still alive"

itchat.auto_login()
itchat.run(blockThread=False)

room = itchat.search_chatrooms("coin")[0]["UserName"]

db = redis.StrictRedis(db=1)

channel = db.pubsub()
channel.subscribe(**{'channel/wechat.msg': lambda x: itchat.send_msg(x["data"].decode(), room)})

channel.run_in_thread(sleep_time=0.001)

while True:
    db.set("status/wechat.alive", '', 25)
    time.sleep(20)
